streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -16,171 +16,256 @@
+                } catch (final RuntimeException e) {
+                    log.error(
+                        "Failed to close {} cleanly. Attempting to close remaining tasks before re-throwing.",
+                        task
+                    );
+                    taskCloseExceptions.put(task.id(), e);
+                }
+                iterator.remove();
+            }
         }
-        if (!addedActiveTasks.isEmpty()) {
-            addNewActiveTasks(addedActiveTasks);
+
+        if (!taskCloseExceptions.isEmpty()) {
+            final Map.Entry<TaskId, RuntimeException> first = taskCloseExceptions.entrySet().iterator().next();
+            throw new RuntimeException(
+                "Unexpected failure to close " + taskCloseExceptions.size() +
+                    " task(s) [" + taskCloseExceptions.keySet() + "]. " +
+                    "First exception (for task " + first.getKey() + ") follows.", first.getValue()
+            );
         }
-        if (!addedStandbyTasks.isEmpty()) {
-            addNewStandbyTasks(addedStandbyTasks);
+
+        if (!activeTasksToCreate.isEmpty()) {
+            taskCreator.createTasks(consumer, activeTasksToCreate).forEach(this::addNewTask);
         }
 
-        // need to clear restore consumer if it was reading standbys but we have active tasks that may need restoring
-        if (!addedActiveTasks.isEmpty() && restoreConsumerAssignedStandbys) {
-            restoreConsumer.unsubscribe();
-            restoreConsumerAssignedStandbys = false;
+        if (!standbyTasksToCreate.isEmpty()) {
+            standbyTaskCreator.createTasks(consumer, standbyTasksToCreate).forEach(this::addNewTask);
         }
 
-        // Pause all the new partitions until the underlying state store is ready for all the active tasks.
-        log.debug("Pausing all active task partitions until the underlying state stores are ready");
-        pausePartitions();
+        builder.addSubscribedTopics(
+            activeTasks.values().stream().flatMap(Collection::stream).collect(Collectors.toList()),
+            logPrefix
+        );
+
+        rebalanceInProgress = false;
     }
 
-    private void resumeSuspended(final Collection<TopicPartition> assignment) {
-        final Set<TaskId> suspendedTasks = partitionsToTaskSet(assignment);
-        suspendedTasks.removeAll(addedActiveTasks.keySet());
+    private void addNewTask(final Task task) {
+        final Task previous = tasks.put(task.id(), task);
+        if (previous != null) {
+            throw new IllegalStateException("Attempted to create a task that we already owned: " + task.id());
+        }
+    }
 
-        log.debug("Suspended tasks to be resumed: {}", suspendedTasks);
+    /**
+     * @throws IllegalStateException If store gets registered after initialized is already finished
+     * @throws StreamsException if the store's change log does not contain the partition
+     */
+    boolean updateNewAndRestoringTasks() {
```

Comment:
nit: this function name can be renamed now, maybe "initializeNewTasks" is good enough.